cpp/src/tsne/bh_kernels.h

```diff
@@ -167,46 +150,38 @@ BoundingBoxKernel(int *restrict startd,
     posxd[NNODES] = (minx + maxx) * 0.5f;
     posyd[NNODES] = (miny + maxy) * 0.5f;
 
-    #pragma unroll
-    for (int a = 0; a < 4; a++)
-      childd[FOUR_NNODES + a] = -1;
+#pragma unroll
+    for (int a = 0; a < 4; a++) childd[FOUR_NNODES + a] = -1;
   }
 }
 
-
 /**
  * Clear some of the state vectors up.
  */
-__global__ __launch_bounds__(1024, 1) void
-ClearKernel1(int *restrict childd,
-             const int FOUR_NNODES,
-             const int FOUR_N)
-{
+__global__ __launch_bounds__(1024, 1) void ClearKernel1(int *restrict childd,
+                                                        const int FOUR_NNODES,
+                                                        const int FOUR_N) {
   const int inc = blockDim.x * gridDim.x;
-  int k = (FOUR_N & -32) + threadIdx.x + blockIdx.x * blockDim.x; 
+  int k = (FOUR_N & -32) + threadIdx.x + blockIdx.x * blockDim.x;
   if (k < FOUR_N) k += inc;
 
-  // iterate over all cells assigned to thread
-  #pragma unroll
-  for (; k < FOUR_NNODES; k += inc)
-    childd[k] = -1;
+// iterate over all cells assigned to thread
+#pragma unroll
+  for (; k < FOUR_NNODES; k += inc) childd[k] = -1;
 }
 
-
 /**
  * Build the actual KD Tree.
  */
-__global__ __launch_bounds__(THREADS2, FACTOR2) void
-TreeBuildingKernel( /* int *restrict errd, */
-                   int *restrict childd,
-                   const float *restrict posxd,
-                   const float *restrict posyd,
-                   const int NNODES,
-                   const int N,
-                   int *restrict maxdepthd,
-                   int *restrict bottomd,
-                   const float *restrict radiusd)
-{
+__global__ __launch_bounds__(
```

Comment:
These types of long kernels usually has very poor performance. Most of the threads end up getting serialized due to warp divergence. Also, there is no way you can control the memory access patterns like coalesced access. These memory access patters have a huge impact on performance.

Please read this book if you haven't read it already;

https://play.google.com/store/books/details?id=Jgx_BAAAQBAJ&gl=us&hl=en-US&source=productsearch&utm_source=HA_Desktop_US&utm_medium=SEM&utm_campaign=PLA&pcampaignid=MKTAD0930BO1&gclid=EAIaIQobChMIppKg_IzT5AIVlI3ICh2CkAaXEAQYAiABEgJJ1vD_BwE&gclsrc=aw.ds

It explains all these details beautifully. 

I'd highly recommend to review these types of code again in TSNE and break them into pieces, and use ml-prims as much as possible because most of ml-prims have already have coalesced access patterns. Also, since there will possible be a lot less warp divergence because of the less if-else blocks in the code.

There are also concepts like kernel fusion that we are planning to address in the future.

Reply:
While I whole-heartedly agree that our ultimate goal is speed-of-light, it's also okay to achieve that goal iteratively, by delivering something quickly that demonstrates a great initial speedup and optimizing it over time to make it even faster. 

cuML's T-SNE is already demonstrating a significant speedup over sklearn and it only stands to get better as it continues to be refined. Other cuML algorithms have been following this approach as well and it aligns with the AMANALAP acronym I've been seeing thrown around recently.

IMO, this should be a github issue, if one hasn't been opened already.

Reply:
It would have aligned with the AMANALAP if the ML prims used it here, right? Offering to use the existing code without rewriting it.

Reply:
For sure, when it’s straightforward to use existing prims directly or decompose a kernel into them.

Daniel was starting with an existing implementation. Completely agree that the kernels like clearing and initializing values are more straightforward and should directly use prims. 

Efficiently building a quad tree is much less straightforward to do on device. Thejaswi and I have wanted to add good quad-tree /kd-tree prims for awhile. For now, I believe Daniel had the CannyLabs version as a starting point.

Reply:
@oyilmaz-nvidia Cool I'll take a look at the book!
But ye as @cjnolet mentioned, I mostly followed the kd-tree in CannyLab's implementation. Since recursion for now is very complicated in a kernel, that's mostly why the kernel looks massive.

But ye over time, I'll see if I can decompose the kernel, but it'll be pretty hard. For now, cuML TSNE runs in 5s for 204,800 by 80 matrices whilst sklearn takes 10676 seconds (3 hours), which is a massive 2000x speedup!

I actually have data showcasing which kernels take the most time. On n = 1,000,000 and p = 100 [took 55s], finding distances takes 50% of the time, and ClearKernel2 takes 26% of the time. Repulsion takes 20% of the time.

On the long kernels, TreeBuilding takes 2% of the time, but Repulsion does take 26%.
Interestingly a small 5 line kernel [ClearKernel2] is eating 26% of all time, so I'll see what I can do to optimize it.